mp2-v1/src/receipt_extraction/leaf.rs

```diff
@@ -0,0 +1,471 @@
+    /// Byte offset for the address from the beginning of a Log
+    add_rel_offset: Target,
+    /// Packed event signature,
+    event_signature: Array<Target, 32>,
+    /// Byte offset from the start of the log to event signature
+    sig_rel_offset: Target,
+    /// The topics for this Log
+    topics: [LogColumn; 3],
+    /// The extra data stored by this Log
+    data: [LogColumn; 2],
+}
+
+/// Contains all the information for a [`Log`] in rlp form
+#[derive(Debug, Clone, Serialize, Deserialize, Copy, PartialEq, Eq)]
+pub struct LogColumn {
+    column_id: Target,
+    /// The byte offset from the beggining of the log to this target
+    rel_byte_offset: Target,
+    /// The length of this topic/data
+    len: Target,
+}
+
+impl LogColumn {
+    /// Convert to an array for metadata digest
+    pub fn to_array(&self) -> [Target; 3] {
+        [self.column_id, self.rel_byte_offset, self.len]
+    }
+
+    /// Assigns a log colum from a [`LogDataInfo`]
+    pub fn assign(&self, pw: &mut PartialWitness<GFp>, data: LogDataInfo) {
+        pw.set_target(self.column_id, F::from_canonical_usize(data.column_id));
+        pw.set_target(
+            self.rel_byte_offset,
+            F::from_canonical_usize(data.rel_byte_offset),
+        );
+        pw.set_target(self.len, F::from_canonical_usize(data.len));
+    }
+}
+
+impl EventWires {
+    /// Convert to an array  for metadata digest
+    pub fn to_vec(&self) -> Vec<Target> {
+        let topics_flat = self
+            .topics
+            .iter()
+            .flat_map(|t| t.to_array())
+            .collect::<Vec<Target>>();
+        let data_flat = self
+            .data
+            .iter()
+            .flat_map(|t| t.to_array())
+            .collect::<Vec<Target>>();
+        let mut out = Vec::new();
+        out.push(self.size);
+        out.extend_from_slice(&self.address.arr);
+        out.push(self.add_rel_offset);
+        out.extend_from_slice(&self.event_signature.arr);
+        out.push(self.sig_rel_offset);
+        out.extend_from_slice(&topics_flat);
+        out.extend_from_slice(&data_flat);
+
+        out
+    }
+
+    pub fn verify_logs_and_extract_values<const NODE_LEN: usize>(
+        &self,
+        b: &mut CBuilder,
+        value: &VectorWire<Target, { PAD_LEN(NODE_LEN) }>,
+        status_offset: Target,
+        relevant_logs_offsets: &VectorWire<Target, MAX_LOGS_PER_TX>,
+    ) -> CurveTarget {
+        let t = b._true();
+        let zero = b.zero();
+        let curve_zero = b.curve_zero();
+        let mut points = Vec::new();
+
+        // Enforce status is true.
+        let status = value.arr.random_access_large_array(b, status_offset);
+        b.connect(status, t.target);
+
+        for log_offset in relevant_logs_offsets.arr.arr {
+            // Extract the address bytes
+            let address_start = b.add(log_offset, self.add_rel_offset);
+
+            let address_bytes = value.arr.extract_array_large::<_, _, 20>(b, address_start);
```

Comment:
Do you think what you're doing here is better than to extract first an array containing only the event and then do extract on this smaller array ?